controller/program/src/instructions/start_rebalance.rs

```diff
@@ -0,0 +1,491 @@
+    &'a [AccountHandle<'acc>],
+    &'a [AccountHandle<'acc>],
+>;
+
+/// Load current instruction index from Instructions sysvar
+#[inline]
+fn load_current_index(data: &[u8]) -> u16 {
+    if data.len() < 2 {
+        return 0;
+    }
+    let last_index = data.len() - 2;
+    u16::from_le_bytes([data[last_index], data[last_index + 1]])
+}
+
+type ParsedInstruction = ([u8; 32], u8, Vec<[u8; 32]>);
+
+/// Load instruction at index from Instructions sysvar
+#[inline]
+fn load_instruction_at(index: usize, data: &[u8]) -> Result<ParsedInstruction, ProgramError> {
+    let mut current = 0;
+    if data.len() < 2 {
+        return Err(INVALID_ACCOUNT_DATA.into());
+    }
+    let num_instructions = u16::from_le_bytes([data[current], data[current + 1]]);
+    current += 2;
+
+    if index >= num_instructions as usize {
+        return Err(BuiltInProgramError::InvalidArgument.into());
+    }
+
+    // Skip to the instruction offset table entry for our index
+    current += index * 2;
+    if current + 2 > data.len() {
+        return Err(INVALID_ACCOUNT_DATA.into());
+    }
+    let start = u16::from_le_bytes([data[current], data[current + 1]]) as usize;
+
+    // Jump to instruction data
+    current = start;
+    if current + 2 > data.len() {
+        return Err(INVALID_ACCOUNT_DATA.into());
+    }
+
+    // Read num_accounts
+    let num_accounts = u16::from_le_bytes([data[current], data[current + 1]]);
+    current += 2;
+
+    // Parse account metas (each is 1 byte meta + 32 bytes pubkey)
+    let accounts_start = current;
+    let accounts_size = num_accounts as usize * 33;
+    if current + accounts_size > data.len() {
+        return Err(INVALID_ACCOUNT_DATA.into());
+    }
+
+    // Extract account pubkeys
+    let mut account_pubkeys = Vec::with_capacity(num_accounts as usize);
+    for i in 0..num_accounts as usize {
+        let acc_start = accounts_start + i * 33 + 1; // +1 to skip meta byte
+        if acc_start + 32 > data.len() {
+            return Err(INVALID_ACCOUNT_DATA.into());
+        }
+        let mut pubkey = [0u8; 32];
+        pubkey.copy_from_slice(&data[acc_start..acc_start + 32]);
+        account_pubkeys.push(pubkey);
+    }
+
+    current += accounts_size;
+
+    // Read program_id (32 bytes)
+    if current + 32 > data.len() {
+        return Err(INVALID_ACCOUNT_DATA.into());
+    }
+    let program_id_start = current;
+    current += 32;
+
+    // Read data_len
+    if current + 2 > data.len() {
+        return Err(INVALID_ACCOUNT_DATA.into());
+    }
+    let data_len = u16::from_le_bytes([data[current], data[current + 1]]);
+    current += 2;
+
+    // Read first byte of instruction data (discriminator)
+    let discm = if data_len > 0 && current < data.len() {
+        data[current]
+    } else {
+        0
+    };
+
+    let mut program_id = [0u8; 32];
+    program_id.copy_from_slice(&data[program_id_start..program_id_start + 32]);
+
+    Ok((program_id, discm, account_pubkeys))
+}
+
```

Comment:
sry i shouldve let you know in advance: use [`jiminy-sysvar-instructions`](https://github.com/igneous-labs/jiminy/tree/master/sysvar/instructions). 

Ah well at least now you now how the Instructions sysvar works